daft/sql/sql.py

```diff
@@ -35,17 +39,31 @@ def sql_expr(sql: str) -> Expression:
 
 
 @PublicAPI
-def sql(sql: str, catalog: SQLCatalog) -> DataFrame:
+def sql(sql: str, catalog: Optional[SQLCatalog] = None) -> DataFrame:
     """Create a DataFrame from an SQL query.
 
     EXPERIMENTAL: This features is early in development and will change.
 
     Args:
         sql (str): SQL query to execute
+        catalog (SQLCatalog, optional): Catalog of tables to use in the query.
+            Defaults to None, in which case a catalog will be built from variables
+            in the callers scope
 
     Returns:
         DataFrame: Dataframe containing the results of the query
     """
+    if catalog is None:
+        try:
+            # Caller is back from func, analytics, annotation
+            caller_frame = inspect.currentframe().f_back.f_back.f_back  # type: ignore
+            caller_vars = {**caller_frame.f_globals, **caller_frame.f_locals}  # type: ignore
+        except AttributeError as exc:
```

Comment:
I think instead of making it use either the scoped variables, or the global context, we should be able to mix and match. I'd prefer having a `register_globals` parameter: `sql(query, catalog, register_globals=False)`

Reply:
yes sorry for the confusion, Having `register_globals` default to `True` is what I intended.

Reply:
Sorry, I think I might have misunderstood that as a signature. Do you mean by default enable supplying a catalog and adding caller-visible variables to it? `register_globals=False` would mean don't activate this new code, essentially?

Reply:
Yes, to clarify, i meant `sql(query, catalog, register_globals=False)` as what the user would invoke if they did not want to register globals, the signature should be 

```py
def sql(sql: str, catalog: Optional[SQLCatalog] = None, register_globals: bool = True) -> DataFrame:
```

once again, sorry for the confusion. 

> register_globals=False would mean don't activate this new code, essentially?

yes that's correct. Which means this should always be invalid 

`daft.sql(query, register_globals=False)`

---
Not required, but we could go the extra step and `@overload` the method to prevent incorrect usage at the type level

I think this would give us the desired results 


```py
from typing import overload, Optional

@overload
def sql(sql: str) -> DataFrame: ...

@overload
def sql(sql: str, catalog: SQLCatalog, register_globals: bool = True) -> DataFrame: ...

@overload
def sql(sql: str, catalog: SQLCatalog = None, register_globals: bool = True) -> DataFrame: ...

def sql(sql: str, catalog: Optional[SQLCatalog] = None, register_globals: bool = True) -> DataFrame:
    # Implementation here
```

- `sql(query)` -> valid
- `sql(query, catalog)` -> valid
- `sql(query, catalog, register_globals=False)` -> valid
- `sql(query, register_globals=False)` -> type error

Reply:
thanks for the clarification - now I see it better. I had thought of it as if the user supplies a catalog they could populate it as needed and wouldn't need the scoped vars - but this way may be more flexible in case a catalog is passed around - though at this point a rule would have to be imposed on precedence (presumably the catalog is higher than the vars).

For this, I'll have to add some features to `SQLCatalog` to add tables and get registered names. I'll look into that as soon as I can, but might be a second 😄